File: ICPC/KAIST/kaist18spring_s.kt
import kotlin.math.*
import java.util.*
import java.math.*

var str = arrayOf("New Customer\n", "Promising\n", "About to Sleep\n", "Hibernating\n", "Lost\n", "Potential Loyalist\n", "Need Attention\n", "About to Leave\n", "Champion\n", "Loyal Customer\n", "Can't Lose Them\n")
var arrMp = arrayOf( 
	intArrayOf(0, 1, 2, 4, 4),
	intArrayOf(5, 5, 2, 3, 4),
	intArrayOf(5, 5, 6, 7, 7),
	intArrayOf(9, 9, 9, 7, 7),
	intArrayOf(8, 9, 9, 7, 10))

fun main(args : Array<String>){
	val br = java.io.BufferedReader(java.io.InputStreamReader(System.`in`), 32768)
  	val bw = java.io.BufferedWriter(java.io.OutputStreamWriter(System.`out`))
	val arrR = br.readLine()!!.trim().split(" ").map{ it.toInt() } 
	val arrF = br.readLine()!!.trim().split(" ").map{ it.toInt() } 
	val q = br.readLine()!!.trim().toInt()
	var freq = TreeMap<String, Int>();
	var las = TreeMap<String, Int>();
	for(i in 1..q){
		var (x, y) = br.readLine()!!.trim().split(" ")
		if(x == "1"){
			las.put(y, i)
			freq.put(y, (freq.get(y) ?: 0) + 1) 
		}
		else{
			if(!freq.contains(y)){
				bw.write("None\n")
				continue
			}
			var f = freq.get(y)!!
			var r = i - las.get(y)!!
			bw.write(str[arrMp[arrF.filter{ it < f }.size][arrR.filter{ it < r }.size]])
		}
	}
	bw.flush()
	bw.close()
}

File: ICPC/KAIST/kaist18spring_s.cpp
#include <bits/stdc++.h>
using namespace std;
using lint = long long;
using pi = array<lint, 2>;
#define sz(v) ((int)(v).size())
#define all(v) (v).begin(), (v).end()
#define cr(v, n) (v).clear(), (v).resize(n);

vector<string> str = {"New Customer", "Promising",			"About to Sleep", "Hibernating",
					  "Lost",		  "Potential Loyalist", "Need Attention", "About to Leave",
					  "Champion",	  "Loyal Customer",		"Can't Lose Them"};
int mp[5][5] = {{0, 1, 2, 4, 4}, {5, 5, 2, 3, 4}, {5, 5, 6, 7, 7}, {9, 9, 9, 7, 7}, {8, 9, 9, 7, 10}};

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);
	vector<int> r(4), f(4);
	for (auto &x : r)
		cin >> x;
	for (auto &x : f)
		cin >> x;
	sort(all(r));
	sort(all(f));
	int q;
	cin >> q;
	map<string, int> fr, ls;
	map<string, int> freq, las;
	for (int i = 1; i <= q; i++) {
		string x, y;
		cin >> x >> y;
		if (x == "1")
			las[y] = i, freq[y]++;
		else if (!freq.count(y))
			cout << "None\n";
		else {
			int cnt1 = lower_bound(all(f), freq[y]) - f.begin();
			int cnt2 = lower_bound(all(r), i - las[y]) - r.begin();
			cout << str[mp[cnt1][cnt2]] << '\n';
		}
	}
}